common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,313 @@
+                    this._putToken.outstandingPutTokens.splice(i, 1);
+                    //
+                    // This was the last outstanding put token.  No point in having a timer around trying to time nothing out.
+                    //
+                    if (this._putToken.outstandingPutTokens.length === 0) {
+                      clearTimeout(this._putToken.timeoutTimer);
+                    }
+                    /*Codes_SRS_NODE_AMQP_CBS_16_013: [The `putToken` method shall call `callback` (if supplied) if the `send` generates an error such that no response from the service will be forthcoming.]*/
+                    outStandingPutTokenInError.putTokenCallback(err);
+                    break;
+                  }
+                }
+              }
+            });
+          }
+        },
+        detaching: {
+          _onEnter: (forwardedCallback, err) => {
+            /*Codes_SRS_NODE_AMQP_CBS_16_008: [`detach` shall detach both sender and receiver links and return the state machine to the `detached` state.]*/
+            this._senderLink.detach();
+            this._receiverLink.detach();
+            this._fsm.transition('detached', forwardedCallback, err);
+          },
+          '*': (callback) => this._fsm.deferUntilTransition('detached')
+        }
+      }
+    });
+  }
+
+  attach(callback: (err?: Error) => void): void {
+    this._fsm.handle('attach', callback);
+  }
+
+  detach(): void {
+    this._fsm.handle('detach');
+  }
+
+  /**
+   * @method             module:azure-iot-amqp-base.Amqp#putToken
+   * @description        Sends a put token operation to the IoT Hub to provide authentication for a device.
+   * @param              audience          The path that describes what is being authenticated.  An example would be
+   *                                       hub.azure-devices.net%2Fdevices%2Fmydevice
+   * @param              token             The actual sas token being used to authenticate the device.  For the most
+   *                                       part the audience is likely to be the sr field of the token.
+   * @param {Function}   putTokenCallback  Called when the put token operation terminates.
+   */
+  putToken(audience: string, token: string, putTokenCallback: (err?: Error) => void): void {
+    /*Codes_SRS_NODE_AMQP_CBS_16_009: [The `putToken` method shall throw a ReferenceError if the `audience` argument is falsy.]*/
+    if (!audience) {
+      throw new ReferenceError('audience cannot be \'' + audience + '\'');
+    }
+
+    /*Codes_SRS_NODE_AMQP_CBS_16_010: [The `putToken` method shall throw a ReferenceError if the `token` argument is falsy.]*/
+    if (!token) {
+      throw new ReferenceError('token cannot be \'' + token + '\'');
+    }
+
+    this._fsm.handle('putToken', audience, token, putTokenCallback);
+  }
+
+  private _removeExpiredPutTokens(): void {
+    const currentTime = Math.round(Date.now() / 1000);
+    let expiredPutTokens: PutTokenOperation[] = [];
```

Comment:
probably. right now it's nicely self-contained though, so not touching it.

---
In reply to: [127619196](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r127619196) [](ancestors = 127619196)